### PR TITLE
Add .NET MAUI Blazor Hybrid frontend app

### LINKED ISSUE #253: Add MAUI Front-End

### PR DESCRIPTION
**Pull request by Eilon**
## Purpose

Adds a .NET MAUI Blazor Hybrid frontend app that re-uses the same Razor Components used in the Blazor Web frontend app. This helps showcase how a .NET MAUI cross-platform app can share the same Azure OpenAI functionality as a web app, and run natively on Windows, macOS, iOS, and Android devices.

Fixes #253

## Does this introduce a breaking change?

[ ] Yes
[x] No

## Pull Request Type

What kind of change does this Pull Request introduce?

[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[x] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ ] Other... Please describe:

## How to Test

1. Ensure all existing repo pre-reqs are installed for .NET 8 and Visual Studio, including the .NET MAUI workload
1. Clone repo
1. Follow the existing steps in the repo to publish the app to Azure
1. Get the URL of the app that was published to Azure Container Apps (something like `https://MY_HOSTED_APP.example.azurecontainerapps.io/`)
1. Update the file `app/maui-blazor/MauiProgram.cs` to use that URL to set the `client.BaseAddress` value (there is a 'TODO' note in the file)
1. Set the .NET MAUI Blazor Hybrid app as the startup project
1. Run the app on your device of choice (Windows Desktop, iPhone/iPad, MacBook, Android phone/tablet, etc.) and you should see a UI that exactly matches the Blazor Web app, except that the app is running locally (API calls still go out to Azure as usual)
1. Profit

## Other Information

This PR conceptually has a few areas of change:

1. Almost all of the Razor components in the `app/frontend` Blazor Web app were moved to a new shared `app/SharedWebComponents` Razor Class Library (RCL)
2. The Blazor Web app now references that shared code
3. There is a new .NET MAUI Blazor Hybrid project that is mostly just the default template code. Almost all of the brand new code in this PR is standard template code. This new project references the same shared RCL to reuse the exact same functionality and implementation
4. The few notable changes in the apps are:
   1. The PDF viewer logic was put behind a service interface so that it could use a web-based viewer in the Blazor Web app, and in the future a native PDF viewer in the .NET MAUI app
   2. A UI rendering bug was fixed in `app/SharedWebComponents/Pages/Chat.razor` so that it displays properly on narrow screens (this was a pre-existing issue in the web app, but was very noticable on narrow mobile devices)
   3. Some try/catch statements were added in the speech API calls

cc @BretJohnson @mattleibow @maddymontaquila @lutzroeder

### CODE REVIEW

File: app/maui-blazor/MauiBlazor.csproj [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+        <!-- Raw Assets (also remove the "Resources\Raw" prefix) -->
+        <MauiAsset Include="Resources\Raw\**" LogicalName="%(RecursiveDir)%(Filename)%(Extension)" />
+    </ItemGroup>
+
+    <ItemGroup>
+        <PackageReference Include="Microsoft.Maui.Controls" Version="$(MauiVersion)" />
+        <PackageReference Include="Microsoft.Maui.Controls.Compatibility" Version="$(MauiVersion)" />
+        <PackageReference Include="Microsoft.AspNetCore.Components.WebView.Maui" Version="$(MauiVersion)" />
+        <PackageReference Include="Microsoft.Extensions.Logging.Debug" Version="8.0.0-rc.2.23479.6" />
```
Comment (Eilon) [inline app/maui-blazor/MauiBlazor.csproj:68]: Oops @BretJohnson we need to fix these versions too. I'll try to find time for this but let me know if you want to take a crack at it.

File: app/SharedWebComponents/Services/IPdfViewer.cs [hunk 2]
```diff
@@ -0,0 +1,8 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+
+namespace SharedWebComponents.Services;
+
+public interface IPdfViewer
+{
+    Task ShowDocument(string name, string url);
```
Comment (IEvangelist) [inline app/SharedWebComponents/Services/IPdfViewer.cs:7]: Let's use the `Async` suffix for consistency. Also, consider the `ValueTask` as the return type to help avoid allocations if possible.

```suggestion
    ValueTask ShowDocumentAsync(string name, string url);
```

File: app/frontend/Services/WebPdfViewer.cs [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+            },
+            new DialogOptions
+            {
+                MaxWidth = MaxWidth.Large,
+                FullWidth = true,
+                CloseButton = true,
+                CloseOnEscapeKey = true
+            });
+
+        return Task.CompletedTask;
```
Comment (IEvangelist) [inline app/frontend/Services/WebPdfViewer.cs:26]:
```suggestion
        return ValueTask.CompletedTask;
```

File: app/frontend/Services/WebPdfViewer.cs [hunk 4]
```diff
@@ -0,0 +1,28 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+
+using ClientApp.Components;
+
+namespace ClientApp.Services;
+
+public class WebPdfViewer(IDialogService dialog) : IPdfViewer
+{
+    public Task ShowDocument(string name, string url)
```
Comment (IEvangelist) [inline app/frontend/Services/WebPdfViewer.cs:9]:
```suggestion
    public ValueTask ShowDocumentAsync(string name, string url)
```

File: app/maui-blazor/MauiBlazor.csproj [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+      <PackageReference Include="MudBlazor" Version="6.11.0" />
```
Comment (IEvangelist) [inline app/maui-blazor/MauiBlazor.csproj:69]: Correct XML indentation.

File: app/maui-blazor/Services/MauiPdfViewer.cs [hunk 6]
```diff
@@ -0,0 +1,11 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+
+namespace MauiBlazor.Services;
+
+public class MauiPdfViewer : IPdfViewer
+{
+    public async Task ShowDocument(string name, string url)
```
Comment (IEvangelist) [inline app/maui-blazor/Services/MauiPdfViewer.cs:7]:
```suggestion
    public async ValueTask ShowDocumentAsync(string name, string url)
```

File: app/maui-blazor/Services/ServiceImplementations.cs [hunk 7]
```diff
@@ -0,0 +1,271 @@
...
+    }
+
+    public void RemoveItem(string key)
+    {
+    }
+
+    public void SetItem<TValue>(string key, TValue value, JsonSerializerOptions? options = null)
+    {
+    }
+}
```
Comment (IEvangelist) [inline app/maui-blazor/Services/ServiceImplementations.cs:37]: Implement this with a `ConcurrentDictionary<string, object>()`. 

```suggestion
public class MauiSessionStorageService : ISessionStorageService
{
    private readonly ConcurrentDictionary<string, object> _storage = new(StringComparison.OrdinalIgnoreCase);

    public double Length => _storage.Length;

    public void Clear() => _storage.Clear();

    public TValue? GetItem<TValue>(string key, JsonSerializerOptions? options = null)
    {
        // Ignore these...
        _ = options;
        
        return _storage.TryGetValue(key, out var value)
            ? value as TValue
            : default;
    }

    public string? Key(double index) => default;

    public void RemoveItem(string key)
    {
        _storage.TryRemove(key, out var _);
    }

    public void SetItem<TValue>(string key, TValue value, JsonSerializerOptions? options = null)
    {
        // Ignore these...
        _ = options;
        
        _storage.AddOrUpdate(key, value, (oldValue, newValue) => newValue);
    }
}
```

**Review by IEvangelist — APPROVED** [review]
This looks awesome! A few minor suggestions... 💜

File: app/maui-blazor/MauiBlazor.csproj (hunk 1 above)
Comment (BretJohnson) [reply app/maui-blazor/MauiBlazor.csproj]: @Eilon - I updated 8.0.0-rc.2.23479.6 -> 8.0.0. What about `$(MauiVersion)` ?

File: app/maui-blazor/MauiBlazor.csproj (hunk 5 above)
Comment (BretJohnson) [reply app/maui-blazor/MauiBlazor.csproj]: Updated

File: app/maui-blazor/MauiBlazor.csproj (hunk 1 above)
Comment (rmarinho) [reply app/maui-blazor/MauiBlazor.csproj]: You should leave that one

File: app/SharedWebComponents/Services/IPdfViewer.cs (hunk 2 above)
Comment (BretJohnson) [reply app/SharedWebComponents/Services/IPdfViewer.cs]: Updated

File: app/frontend/Services/WebPdfViewer.cs (hunk 3 above)
Comment (BretJohnson) [reply app/frontend/Services/WebPdfViewer.cs]: Updated

File: app/frontend/Services/WebPdfViewer.cs (hunk 4 above)
Comment (BretJohnson) [reply app/frontend/Services/WebPdfViewer.cs]: Updated

File: app/maui-blazor/Services/MauiPdfViewer.cs (hunk 6 above)
Comment (BretJohnson) [reply app/maui-blazor/Services/MauiPdfViewer.cs]: Updated

File: app/SharedWebComponents/Components/Answer.razor.cs [hunk 8]
```diff
@@ -28,7 +28,7 @@ private async Task OnAskFollowupAsync(string followupQuestion)
             await FollowupQuestionClicked.InvokeAsync(followupQuestion);
         }
     }
-    private void OnShowCitation(CitationDetails citation) => PdfViewer.ShowDocument(citation.Name, citation.BaseUrl);
+    private void OnShowCitation(CitationDetails citation) => PdfViewer.ShowDocumentAsync(citation.Name, citation.BaseUrl);
```
Comment (IEvangelist) [inline app/SharedWebComponents/Components/Answer.razor.cs:31]:
```suggestion
    private Task OnShowCitation(CitationDetails citation) => PdfViewer.ShowDocumentAsync(citation.Name, citation.BaseUrl);
```

File: app/SharedWebComponents/Pages/Docs.razor.cs [hunk 9]
```diff
@@ -101,7 +101,7 @@ private async Task SubmitFilesForUploadAsync()
         }
     }
 
-    private void OnShowDocument(DocumentResponse document) => PdfViewer.ShowDocument(document.Name, document.Url.ToString().Replace($"/{document.Name}", ""));
+    private void OnShowDocument(DocumentResponse document) => PdfViewer.ShowDocumentAsync(document.Name, document.Url.ToString().Replace($"/{document.Name}", ""));
```
Comment (IEvangelist) [inline app/SharedWebComponents/Pages/Docs.razor.cs:104]:
```suggestion
    private Task OnShowDocument(DocumentResponse document) => PdfViewer.ShowDocumentAsync(document.Name, document.Url.ToString().Replace($"/{document.Name}", ""));
```

**Review by IEvangelist — COMMENTED** [review]
These should be `Task` returning.

File: README.md [hunk 10]
```diff
@@ -58,6 +58,8 @@ This sample demonstrates a few approaches for creating ChatGPT-like experiences
 
 The repo includes sample data so it's ready to try end-to-end. In this sample application, we use a fictitious company called Contoso Electronics, and the experience allows its employees to ask questions about the benefits, internal policies, as well as job descriptions and roles.
 
+This fork adds a MAUI Blazor app.
```
Comment (mattleibow) [inline README.md:61]: We can remove this?

File: app/Directory.Packages.props [hunk 11]
```diff
@@ -1,6 +1,7 @@
 <Project>
   <PropertyGroup>
     <ManagePackageVersionsCentrally>true</ManagePackageVersionsCentrally>
+    <MauiVersion>8.0.6</MauiVersion>
```
Comment (mattleibow) [inline app/Directory.Packages.props:4]: Maybe we can just use normal version style and not a property?

Comment (BretJohnson) [reply app/Directory.Packages.props:4]: `MauiVersion` is used 3 times below, so (unless I'm missing something) a property seems better, right?

File: README.md (hunk 10 above)
Comment (BretJohnson) [reply README.md]: Updated

**Comment by BretJohnson** [comment]
@IEvangelist - We think this is all good now, and have marked it non-draft. Can you please review again and merge when you're comfortable. Also, please review my README text updates - it's not bad, but you may have suggested improvements.

File: app/frontend/Services/WebPdfViewer.cs [hunk 12]
```diff
@@ -0,0 +1,28 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+
+using ClientApp.Components;
```
Comment (IEvangelist) [inline app/frontend/Services/WebPdfViewer.cs:3]: This should be removed from here, and instead defined as part of `GlobalUsings`.

File: app/maui-blazor/MauiProgram.cs [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+
+#if DEBUG
+		builder.Services.AddBlazorWebViewDeveloperTools();
+		builder.Logging.AddDebug();
+#endif
+
+		builder.Services.AddHttpClient<ApiClient>(client =>
+		{
+			// TODO: Configure this to point to your deployed API. For example, https://MY_HOSTED_APP.example.azurecontainerapps.io/
+			client.BaseAddress = new Uri("TODO");
```
Comment (IEvangelist) [inline app/maui-blazor/MauiProgram.cs:30]: Is this `TODO` going to be addressed before we approve and merge? Is this something that can be done easily, how is that usually handled with .NET MAUI apps?

File: app/maui-blazor/MauiProgram.cs [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+		builder.Services.AddSingleton<ISessionStorageService, MauiSessionStorageService>();
+		builder.Services.AddSingleton<ISpeechRecognitionService, MauiSpeechRecognitionService>();
+		builder.Services.AddSingleton<ISpeechSynthesisService, MauiSpeechSynthesisService>();
+		builder.Services.AddTransient<IPdfViewer, MauiPdfViewer>();
+
+		builder.Services.AddMudServices();
+
+		builder.Services.AddSingleton<IPreferences>(Preferences.Default);
+		builder.Services.AddSingleton<ISpeechToText>(SpeechToText.Default);
+
```
Comment (IEvangelist) [inline app/maui-blazor/MauiProgram.cs:44]: I would have expected to see:

```csharp
builder.Services.AddSingleton<ITextToSpeech>(TextToSpeech.Default);
```

Here as well, I think we should add both features.

File: app/maui-blazor/Platforms/MacCatalyst/AppDelegate.cs [hunk 15]
```diff
@@ -0,0 +1,9 @@
+﻿using Foundation;
+
+namespace MauiBlazor;
+
+[Register("AppDelegate")]
```
Comment (IEvangelist) [inline app/maui-blazor/Platforms/MacCatalyst/AppDelegate.cs:5]: Should this use `nameof`?

```suggestion
[Register(nameof(AppDelegate))]
```

File: app/maui-blazor/Services/ServiceImplementations.cs [hunk 16]
```diff
@@ -0,0 +1,279 @@
...
+            }
+        }
+
+        var options = new SpeechOptions
+        {
+            Locale = locale
+        };
+        await TextToSpeech.Default.SpeakAsync(utterance.Text, options, _cts.Token);
+
+        _cts = null;
```
Comment (IEvangelist) [inline app/maui-blazor/Services/ServiceImplementations.cs:114]: I think this is wrong. If you `null` it out, there's no way to cancel later as you no longer have a reference to the corresponding Token passed into the `SpeakAsync` API above.

File: app/maui-blazor/Services/ServiceImplementations.cs [hunk 17]
```diff
@@ -0,0 +1,279 @@
...
+}
+
+public class MauiLocalStorageService : ILocalStorageService
+{
+    private readonly IPreferences _prefs;
+
+    public MauiLocalStorageService(IPreferences prefs)
+    {
+        _prefs = prefs;
+    }
```
Comment (IEvangelist) [inline app/maui-blazor/Services/ServiceImplementations.cs:125]:
```suggestion
public class MauiLocalStorageService(IPreferences prefs) : ILocalStorageService
{
```

File: app/maui-blazor/Services/ServiceImplementations.cs [hunk 18]
```diff
@@ -0,0 +1,279 @@
...
+
+    public double Length => 0;
+
+    public void Clear() =>
+        _prefs.Clear();
```
Comment (IEvangelist) [inline app/maui-blazor/Services/ServiceImplementations.cs:130]:
```suggestion
    public void Clear() => prefs.Clear();
```

File: app/maui-blazor/Services/ServiceImplementations.cs [hunk 19]
```diff
@@ -0,0 +1,279 @@
...
+
+    public TValue? GetItem<TValue>(string key, JsonSerializerOptions? options = null)
+    {
+        // if the type is a nullable, then use the underlying type for parsing
+        if (Nullable.GetUnderlyingType(typeof(TValue)) is Type under)
+        {
+            var get = _prefs.GetType().GetMethod("Get")!;
```
Comment (IEvangelist) [inline app/maui-blazor/Services/ServiceImplementations.cs:137]:
```suggestion
            var get = prefs.GetType().GetMethod("Get")!;
```

File: app/maui-blazor/Services/ServiceImplementations.cs [hunk 20]
```diff
@@ -0,0 +1,279 @@
...
+            var args = new object?[] { key, Activator.CreateInstance(under), default(string) };
+            return (TValue?)get.MakeGenericMethod(under).Invoke(_prefs, args);
+        }
+        else
+        {
+#pragma warning disable CS8604 // Possible null reference argument.
+            return _prefs.Get<TValue>(key, default);
```
Comment (IEvangelist) [inline app/maui-blazor/Services/ServiceImplementations.cs:144]:
```suggestion
            return prefs.Get<TValue>(key, default);
```

File: app/maui-blazor/Services/ServiceImplementations.cs [hunk 21]
```diff
@@ -0,0 +1,279 @@
...
+#pragma warning restore CS8604 // Possible null reference argument.
+        }
+    }
+
+    public string? Key(double index) => default;
+
+    public void RemoveItem(string key) =>
+        _prefs.Remove(key);
```
Comment (IEvangelist) [inline app/maui-blazor/Services/ServiceImplementations.cs:152]:
```suggestion
    public void RemoveItem(string key) => prefs.Remove(key);
```

File: app/maui-blazor/Services/ServiceImplementations.cs [hunk 22]
```diff
@@ -0,0 +1,279 @@
...
+
+    public void SetItem<TValue>(string key, TValue value, JsonSerializerOptions? options = null)
+    {
+        // if the type is a nullable, then use the underlying type for parsing
+        if (Nullable.GetUnderlyingType(typeof(TValue)) is Type under)
+        {
+            var set = _prefs.GetType().GetMethod("Set")!;
```
Comment (IEvangelist) [inline app/maui-blazor/Services/ServiceImplementations.cs:159]:
```suggestion
            var set = prefs.GetType().GetMethod("Set")!;
```

File: app/maui-blazor/Services/ServiceImplementations.cs [hunk 23]
```diff
@@ -0,0 +1,279 @@
...
+        _prefs.Clear();
+
+    public TValue? GetItem<TValue>(string key, JsonSerializerOptions? options = null)
+    {
+        // if the type is a nullable, then use the underlying type for parsing
+        if (Nullable.GetUnderlyingType(typeof(TValue)) is Type under)
+        {
+            var get = _prefs.GetType().GetMethod("Get")!;
+            var args = new object?[] { key, Activator.CreateInstance(under), default(string) };
+            return (TValue?)get.MakeGenericMethod(under).Invoke(_prefs, args);
```
Comment (IEvangelist) [inline app/maui-blazor/Services/ServiceImplementations.cs:139]:
```suggestion
            return (TValue?)get.MakeGenericMethod(under).Invoke(prefs, args);
```

File: app/maui-blazor/Services/ServiceImplementations.cs [hunk 24]
```diff
@@ -0,0 +1,279 @@
...
+        _prefs.Remove(key);
+
+    public void SetItem<TValue>(string key, TValue value, JsonSerializerOptions? options = null)
+    {
+        // if the type is a nullable, then use the underlying type for parsing
+        if (Nullable.GetUnderlyingType(typeof(TValue)) is Type under)
+        {
+            var set = _prefs.GetType().GetMethod("Set")!;
+            var args = new object?[] { key, value, default(string) };
+            set.MakeGenericMethod(under).Invoke(_prefs, args);
```
Comment (IEvangelist) [inline app/maui-blazor/Services/ServiceImplementations.cs:161]:
```suggestion
            set.MakeGenericMethod(under).Invoke(prefs, args);
```

File: app/maui-blazor/Services/ServiceImplementations.cs [hunk 25]
```diff
@@ -0,0 +1,279 @@
...
+        }
+        else
+        {
+            _prefs.Set<TValue>(key, value);
```
Comment (IEvangelist) [inline app/maui-blazor/Services/ServiceImplementations.cs:165]:
```suggestion
            prefs.Set<TValue>(key, value);
```

File: app/maui-blazor/Services/ServiceImplementations.cs [hunk 26]
```diff
@@ -0,0 +1,279 @@
...
+public class MauiSpeechRecognitionService : ISpeechRecognitionService
+{
+    private readonly ISpeechToText _speechToText;
+
+    private SpeechRecognitionOperation? _current;
+
+    public MauiSpeechRecognitionService(ISpeechToText speechToText)
+    {
+        _speechToText = speechToText;
+    }
```
Comment (IEvangelist) [inline app/maui-blazor/Services/ServiceImplementations.cs:179]:
```suggestion
public class MauiSpeechRecognitionService(ISpeechToText speechToText) : ISpeechRecognitionService
{
    private SpeechRecognitionOperation? _current;
```

File: app/maui-blazor/Services/ServiceImplementations.cs [hunk 27]
```diff
@@ -0,0 +1,279 @@
...
+
+    public Task InitializeModuleAsync(bool logModuleDetails = true)
+    {
+        return Task.CompletedTask;
+    }
+
+    public IDisposable RecognizeSpeech(string language, Action<string> onRecognized, Action<SpeechRecognitionErrorEvent>? onError = null, Action? onStarted = null, Action? onEnded = null)
+    {
+        return _current = new SpeechRecognitionOperation(_speechToText, language, onRecognized, onError, onStarted, onEnded);
```
Comment (IEvangelist) [inline app/maui-blazor/Services/ServiceImplementations.cs:199]:
```suggestion
        return _current = new SpeechRecognitionOperation(speechToText, language, onRecognized, onError, onStarted, onEnded);
```

File: app/maui-blazor/Services/ServiceImplementations.cs
Comment (IEvangelist) [inline app/maui-blazor/Services/ServiceImplementations.cs:1]: I would recommend separating each implementation into its own file, these look great BTW!

**Review by IEvangelist — APPROVED** [review]
This looks good, thank you. We'll :shipit: when you make the suggested changes.

File: app/SharedWebComponents/Components/VoiceDialog.razor.cs [hunk 28]
```diff
@@ -21,7 +21,16 @@ protected override async Task OnInitializedAsync()
...
         await GetVoicesAsync();
-        SpeechSynthesis.OnVoicesChanged(() => GetVoicesAsync(true));
+
+        try
+        {
+            SpeechSynthesis.OnVoicesChanged(() => GetVoicesAsync(true));
+        }
+        catch
+        {
+            // do nothing if the service does not support observation
```
Comment (IEvangelist) [inline app/SharedWebComponents/Components/VoiceDialog.razor.cs:31]: I'm not sure I follow this comment, what's the concern this addresses?

File: app/SharedWebComponents/Components/VoiceDialog.razor.cs [hunk 29]
```diff
@@ -55,7 +64,18 @@ private async Task GetVoicesAsync(bool isFromCallback = false)
...
-    public void Dispose() => SpeechSynthesis.UnsubscribeFromVoicesChanged();
+    public void Dispose()
+    {
+        try
+        {
+            SpeechSynthesis.UnsubscribeFromVoicesChanged();
+        }
+        catch
+        {
+            // do nothing if the service does not support observation
```
Comment (IEvangelist) [inline app/SharedWebComponents/Components/VoiceDialog.razor.cs:75]: Same question as above... what's the concern this addresses?

File: app/SharedWebComponents/Pages/VoiceChat.razor.cs [hunk 30]
```diff
@@ -77,11 +79,24 @@ private void OnSendPrompt()
...
+                        {
+                            SpeechSynthesis.Speak(utterance, duration =>
+                            {
+                                _isReadingResponse = false;
+                                StateHasChanged();
+                            });
+                        }
+                        catch
                         {
+                            // TODO: try and not use the method above as it is not really useful
```
Comment (IEvangelist) [inline app/SharedWebComponents/Pages/VoiceChat.razor.cs:92]: Can you explain this to me in a bit more detail? I'm trying to understand the concerns, the `catch`, what "specific type" is referring to.

File: app/maui-blazor/MauiProgram.cs (hunk 13 above)
Comment (Eilon) [reply app/maui-blazor/MauiProgram.cs:30]: For now this is something that the person running the app needs to do. I think it's mentioned in the README comments. This URL needs to be set to wherever the app is published on Azure. There's no automatic way to do that _yet_ but we're looking at possible product features that would enable this automatically in the future (perhaps you saw some earlier thinkings on this by having something like `azd up` set it automatically).

File: app/maui-blazor/Platforms/MacCatalyst/AppDelegate.cs (hunk 15 above)
Comment (Eilon) [reply app/maui-blazor/Platforms/MacCatalyst/AppDelegate.cs:5]: I'm not sure but these might not be semantically the same even though the text is the same. It's from the default MAUI project template so I think it's best to preserve it.

**Comment by mattleibow** [comment]
I believe I got all the feedback items... I updated the comments to be more descriptive too.

File: README.md [hunk 31]
```diff
@@ -245,6 +245,20 @@ If you have existing resources in Azure that you wish to use, you can configure
...
+or on Android and iOS devices. The MAUI client here is implemented using Blazor hybrid, letting it share most
+code with the website frontend.
+
+1. Open `app\app-maui.sln` to open the solution that includes the MAUI client
+
+1. Edit `app\maui-blazor\MauiProgram.cs`, updating `client.BaseAddress` with the URL for the backend.
+If it's running in Azure, use the URL for the service backend from the steps above. If running locally, use
+http://localhost:7181.
+
+1. Set MauiBlazor as the startup project and run the app
```
[the remark this answers is not in the document]
Comment (mattleibow) [reply README.md]: Fixed.

File: app/SharedWebComponents/Components/Answer.razor.cs [hunk 32]
```diff
@@ -1,13 +1,15 @@
 ﻿// Copyright (c) Microsoft. All rights reserved.
 
-namespace ClientApp.Components;
+using System.Reflection.Metadata;
```
[the remark this answers is not in the document]
Comment (mattleibow) [reply app/SharedWebComponents/Components/Answer.razor.cs]: Fixed.

File: app/frontend/Services/WebPdfViewer.cs (hunk 12 above)
Comment (mattleibow) [reply app/frontend/Services/WebPdfViewer.cs]: Fixed.

File: app/maui-blazor/MauiProgram.cs (hunk 13 above)
Comment (mattleibow) [reply app/maui-blazor/MauiProgram.cs:28]: Added a better comment

File: app/maui-blazor/MauiProgram.cs (hunk 14 above)
Comment (mattleibow) [reply app/maui-blazor/MauiProgram.cs:43]: Fixed.

File: app/maui-blazor/Platforms/MacCatalyst/AppDelegate.cs (hunk 15 above)
Comment (mattleibow) [reply app/maui-blazor/Platforms/MacCatalyst/AppDelegate.cs]: Fixed.

File: app/maui-blazor/Services/ServiceImplementations.cs (hunk 16 above)
Comment (mattleibow) [reply app/maui-blazor/Services/ServiceImplementations.cs]: Fixed.

File: app/maui-blazor/Services/ServiceImplementations.cs (hunk 17 above)
Comment (mattleibow) [reply app/maui-blazor/Services/ServiceImplementations.cs]: Fixed.

File: app/maui-blazor/Services/ServiceImplementations.cs (hunk 18 above)
Comment (mattleibow) [reply app/maui-blazor/Services/ServiceImplementations.cs]: Fixed.

File: app/maui-blazor/Services/ServiceImplementations.cs (hunk 20 above)
Comment (mattleibow) [reply app/maui-blazor/Services/ServiceImplementations.cs]: Fixed.

File: app/maui-blazor/Services/ServiceImplementations.cs (hunk 21 above)
Comment (mattleibow) [reply app/maui-blazor/Services/ServiceImplementations.cs]: Fixed.

File: app/maui-blazor/Services/ServiceImplementations.cs (hunk 22 above)
Comment (mattleibow) [reply app/maui-blazor/Services/ServiceImplementations.cs]: Fixed.

File: app/maui-blazor/Services/ServiceImplementations.cs (hunk 23 above)
Comment (mattleibow) [reply app/maui-blazor/Services/ServiceImplementations.cs]: Fixed.

File: app/maui-blazor/Services/ServiceImplementations.cs (hunk 24 above)
Comment (mattleibow) [reply app/maui-blazor/Services/ServiceImplementations.cs]: Fixed.

File: app/maui-blazor/Services/ServiceImplementations.cs (hunk 25 above)
Comment (mattleibow) [reply app/maui-blazor/Services/ServiceImplementations.cs]: Fixed.

File: app/maui-blazor/Services/ServiceImplementations.cs (hunk 26 above)
Comment (mattleibow) [reply app/maui-blazor/Services/ServiceImplementations.cs]: Fixed.

File: app/maui-blazor/Services/ServiceImplementations.cs (hunk 27 above)
Comment (mattleibow) [reply app/maui-blazor/Services/ServiceImplementations.cs]: Fixed.

File: app/SharedWebComponents/Components/VoiceDialog.razor.cs (hunk 28 above)
Comment (mattleibow) [reply app/SharedWebComponents/Components/VoiceDialog.razor.cs]: Added a much better comment.

File: app/SharedWebComponents/Components/VoiceDialog.razor.cs (hunk 29 above)
Comment (mattleibow) [reply app/SharedWebComponents/Components/VoiceDialog.razor.cs]: Added a much better comment.

File: app/SharedWebComponents/Pages/VoiceChat.razor.cs (hunk 30 above)
Comment (mattleibow) [reply app/SharedWebComponents/Pages/VoiceChat.razor.cs]: Added a much better comment.

File: app/maui-blazor/Services/ServiceImplementations.cs
Comment (mattleibow) [reply app/maui-blazor/Services/ServiceImplementations.cs:1]: Good idea. Done!

**Comment by BretJohnson** [comment]
@IEvangelist - we believe we've addressed everything; can you take another look and :shipit: if you think we're ready

**Review by IEvangelist — APPROVED** [review]

**Comment by mattleibow** [comment]
Exciting stuff!

**Pull request merged**